launcher/src/main/java/org/apache/spark/launcher/AbstractLauncher.java

```diff
@@ -0,0 +1,600 @@
+    }
+    return opts;
+  }
+
+  private void finishOpt(List<String> opts, StringBuilder opt) {
+    opts.add(opt.toString());
+    opt.setLength(0);
+  }
+
+  private String findAssembly(String scalaVersion) {
+    String sparkHome = getSparkHome();
+    File libdir;
+    if (new File(sparkHome, "RELEASE").isFile()) {
+      libdir = new File(sparkHome, "lib");
+      checkState(libdir.isDirectory(), "Library directory '%s' does not exist.",
+          libdir.getAbsolutePath());
+    } else {
+      libdir = new File(sparkHome, String.format("assembly/target/scala-%s", scalaVersion));
+    }
+
+    final Pattern re = Pattern.compile("spark-assembly.*hadoop.*\\.jar");
+    FileFilter filter = new FileFilter() {
+      @Override
+      public boolean accept(File file) {
+        return file.isFile() && re.matcher(file.getName()).matches();
+      }
+    };
+    File[] assemblies = libdir.listFiles(filter);
+    checkState(assemblies != null && assemblies.length > 0, "No assemblies found in '%s'.", libdir);
+    checkState(assemblies.length == 1, "Multiple assemblies found in '%s'.", libdir);
+    return assemblies[0].getAbsolutePath();
+  }
+
+  private String getenv(String key) {
+    return first(env != null ? env.get(key) : null, System.getenv(key));
+  }
+
+  /**
+   * Prepare a command line for execution on Windows.
+   *
+   * Two things need to be done:
+   *
+   * - If a custom library path is needed, extend PATH to add it. Based on:
+   *   http://superuser.com/questions/223104/setting-environment-variable-for-just-one-command-in-windows-cmd-exe
+   *
+   * - Quote all arguments so that spaces are handled as expected. Quotes within arguments are
+   *   "double quoted" (which is batch for escaping a quote). This page has more details about
+   *   quoting and other batch script fun stuff: http://ss64.com/nt/syntax-esc.html
+   */
+  private List<String> prepareForWindows(List<String> cmd,
+      String libPath,
+      Map<String, String> env) {
+    StringBuilder cmdline = new StringBuilder("cmd /c \"");
+    if (libPath != null) {
+      cmdline.append("set PATH=%PATH%;").append(libPath).append(" &&");
+    }
+    for (Map.Entry<String, String> e : env.entrySet()) {
+      cmdline.append(String.format("set %s=%s", e.getKey(), e.getValue()));
+      cmdline.append(" &&");
+    }
+    for (String arg : cmd) {
+      if (cmdline.length() > 0) {
+        cmdline.append(" ");
+      }
+      cmdline.append(quote(arg));
+    }
+    cmdline.append("\"");
+    return Arrays.asList(cmdline.toString());
+  }
+
+  /**
+   * Quoting arguments that don't need quoting in Windows seems to cause weird issues. So only
+   * quote arguments when there is whitespace in them.
+   */
+  private boolean needsQuoting(String arg) {
+    for (int i = 0; i < arg.length(); i++) {
+      if (Character.isWhitespace(arg.codePointAt(i))) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private String quote(String arg) {
```

Comment:
I'd rather keep methods only used in a particular class together with that class's code.